Convert CHANGELOG.next to fragments

endorama opened the pull request:
This PR converts the current CHANGELOG.next into fragments that can be leveraged by `elastic-agent-chagenlog-tool`.

I decided to introduce some changes in the actual CHANGELOG to accomodate some missing feature of the tool by removing some data, that are available in any case from the referenced PRs.

We don't plan to backport this.

cc @cmacknz

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-1244/detail/PR-1244/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-1244/detail/PR-1244/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-1244/detail/PR-1244/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-1244/detail/PR-1244/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2022-09-20T09:41:15.398Z&rangeTo=2022-09-20T10:01:15.398Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=cce519262887804525138c06b7bb10bd&transactionId=23afb22927562a60)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-09-20T09:51:15.398+0000


* Duration: 3 min 16 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->
















## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 97.333% (`73/75`) | :+1: 
Files | 68.085% (`160/235`) | :+1: 
Classes | 68.478% (`315/460`) | :+1: 
Methods | 52.826% (`888/1681`) | :+1: 
Lines | 38.884% (`9557/24578`) | :+1: 0.033
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

endorama commented:
/test

cmacknz commented:
I don't see a version in any of the changelog fragments, how do we know which release a fragment applies to?

If the fragments are only meant to exist for the CHANGELOG.next, one issue with the agent changelog is we haven't been maintaining properly the way it would be maintained for Beats. Specifically entries have not been moved from CHANGELOG.next to CHANGELOG after release. 

CHANGELOG.next includes many entries for released changes. The only way to know what the true CHANGELOG.next should be is by looking at what has already been in the previous release notes: https://www.elastic.co/guide/en/fleet/current/release-notes.html

I suspect we need to move many of the existing changelog entry so that we only generate fragments for the "true" 8.5.0 changelog.

endorama commented:
> I don't see a version in any of the changelog fragments, how do we know which release a fragment applies to?

The version is determined when consolidating the changelog. Fragments are version agnostic, the changelog is version aware. This is a PR that should showcase where the version fits in: https://github.com/elastic/elastic-agent-changelog-tool/pull/106 

> CHANGELOG.next includes many entries for released changes. [...] I suspect we need to move many of the existing changelog entry so that we only generate fragments for the "true" 8.5.0 changelog. I can attempt this with a separate PR.

I didn't know that, but I think the easiest solution would be to remove all unnecessary fragments. From my understanding it should suffice matching the entries in the release notes with the `summary` in the fragments.

endorama commented:
By matching release note lines with fragments I found only 1 entry duplicated (from 8.4.0 release notes) and removed in https://github.com/elastic/elastic-agent/pull/1244/commits/d9c729b1ec85eab2dab07fd4401f2b4ddc40d9ff

AndersonQ reviewed (approved):
It's hard to tell that every fragment is correct, but it seems so :)

endorama commented:
A note on the number of files: there are [70 commits](https://github.com/elastic/elastic-agent/compare/v8.4.0...main) from 8.4.0 to current `main`. I don't know how many of this have a changelog entry, but 187 feels a bit too much. 

I'll look into beats release notes to see if there are leftovers from there.

cmacknz commented:
Thanks for looking through these @endorama! 

The Github branch comparison is a nice way to view CHANGELOG entries that are in 8.5 but not 8.4: https://github.com/elastic/elastic-agent/compare/8.4...8.5

We only need to keep changelog entries in 8.5 that aren't already in the 8.4 changelog, which means we can remove any of the commits above that have been backported to 8.4. 

From your list above we can remove these backported changes:
* https://github.com/elastic/elastic-agent/commit/a0620af9818d36353241a518d00ba34ab2ff9d11 Use at least warning level for all status logs (https://github.com/elastic/elastic-agent/pull/1218)
* https://github.com/elastic/elastic-agent/commit/56787a61c0b7faed1114160e01dfaa812cd1a95d Fix/panic with composable renderer (https://github.com/elastic/elastic-agent/pull/823)
* https://github.com/elastic/elastic-agent/commit/6e2e06c06743fde7cb4fbb06934c720d59f9959a Remove the fleet reporter (https://github.com/elastic/elastic-agent/pull/1130)
* https://github.com/elastic/elastic-agent/commit/2522be42faa5f3ec8b562ba7a79b8646d4b160c8 Fix lookup issues with inputs.d fragment yml (https://github.com/elastic/elastic-agent/pull/840)

We can keep:
* https://github.com/elastic/elastic-agent/commit/a0a3ed1aae5436b720b8065199afc629fe8930d6 Add Go 1.18 upgrade to breaking changes section. (https://github.com/elastic/elastic-agent/pull/1216)
* https://github.com/elastic/elastic-agent/commit/c0367b97a255484c3414995f32cefe6de3a12f17 Add Filebeat lumberjack input to spec (https://github.com/elastic/elastic-agent/pull/959)
* https://github.com/elastic/elastic-agent/commit/59f9ac46ec61208005dc2b9806e2fbe10af36c40 Add support for hints' based autodiscovery in kubernetes provider (https://github.com/elastic/elastic-agent/pull/698)

The changelog for 8.5.0 is small, because the majority of bug fixes are being backported to 8.4, and the major feature work we did to support the agent v2 architecture was deferred to 8.6.

If we were maintained the changelog properly the backported entries would be removed when the 8.4.x changelogs were forward ported back to main after release. We do this in Beats but not in Agent.

cmacknz commented on CHANGELOG.next.asciidoc:
```diff
@@ -104,17 +103,17 @@
 - Add fleet-server to output of elastic-agent inspect output command (and diagnostic bundle). {pull}243[243]
 - Update API calls that the agent makes to Kibana when running the container command. {pull}253[253]
-- diagnostics collect log names are fixed on Windows machines, command will ignore failures. AgentID is included in diagnostics(and diagnostics collect) output. {issue}81[81] {issue}92[92] {issue}190[190] {pull}262[262]
+- diagnostics collect log names are fixed on Windows machines, command will ignore failures. AgentID is included in diagnostics(and diagnostics collect) output. {pull}262[262]
```
Once we generate the correct set of fragment files, is there a need to keep the changelog.next file at all? Can we just delete it as part of this PR?

endorama replied:
yes!

cmacknz commented:
Adding the 8.5 backport label since the branch was created before we merged this change.

cmacknz reviewed (changes requested):
Requesting changes since there are some approvals and we still need to tidy up the changelog a bit.

@endorama we have some new changes in main that have been backported to 8.5 now that we'll have to update to include here as well.

endorama commented:
@cmacknz so this PR targets `main`, while from my understanding should now target 8.5? I rebased on top of 8.5, let me know if this was not expected.

I removed all fragments and left only the ones that should be part of this release, may you please double check?

cmacknz commented on CHANGELOG.next.asciidoc:
```diff
@@ -64,7 +64,6 @@
 - Fixed: limit for retries to Kibana configurable {issue-beats}[25063]
```
Let's delete this entire file since we don't need it anymore. That will be a good hint to use the changelog tool instead.

cmacknz commented on README.md:
```diff
@@ -10,7 +10,7 @@ in the [observability-docs](https://github.com/elastic/observability-docs) repo.
 Prerequisites:
```
Can you add a link to the changelog tool in the README?

cmacknz commented:
There are changes in this PR other than the changelog that we need to remove (probably just merging or rebasing with main will fix it). Then:

1) Let's remove the changelog.next file
2) Let's add a link to the changelog tool to the agent README file.
3) Once you update against main, you will need to generate a changelog file for https://github.com/elastic/elastic-agent/pulls which appears to be the only new change since the last update.
4) Merge this to main and we'll backport it to 8.5. Then all releases from 8.5 onward will be using the new changelog process.

endorama commented:
> 1. Let's remove the changelog.next file
> 2. Let's add a link to the changelog tool to the agent README file.

Done

> 3. Once you update against main, you will need to generate a changelog file for [elastic/elastic-agent/pulls](https://github.com/elastic/elastic-agent/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc) which appears to be the only new change since the last update.

The link brings to the list of all open PRs, I guess that was not intended?

cmacknz commented:
> Once you update against main, you will need to generate a changelog file for [elastic/elastic-agent/pulls](https://github.com/elastic/elastic-agent/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc) which appears to be the only new change since the last update.
The link brings to the list of all open PRs, I guess that was not intended?

🤦 No, not intended. The link in the changelog.next file I copied is actually wrong, it should link to https://github.com/elastic/elastic-agent/pull/1369

Regardless I think this looks good enough to merge, I can do another pass over the commits in 8.5 before release to double check we got everything.

endorama commented:
:tada: great! Thank you.
